File: leetcode/647-palindromic-substrings/palindromic-substrings.js
const countSubstrings = s => {
    let count = 0;

    for (let i = 0; i < s.length; i++) {
        // cases like bab
        for (let j = 0; i - j >= 0 && j + i < s.length && s[i - j] === s[j + i]; j++) count++;
        // cases like baab
        for (let j = 0; i - 1 - j >= 0 && j + i < s.length && s[i - 1 - j] === s[j + i]; j++) count++;
    }

    return count;
};

File: leetcode/647-palindromic-substrings/PalindromicSubstrings.java
class Solution {
    int count = 0;
    
    private void scan(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            count++;
            left--;
            right++;
        }
    }
    
    public int countSubstrings(String s) {
        for (int i = 0; i < s.length(); i++) {
            scan(s, i, i);
            scan(s, i, i + 1);
        }
        
        return count;
    }
}
